runtime/lua/vim/provider/python.lua

```diff
@@ -0,0 +1,125 @@
+local M = {}
+
+--- @param prog string
+--- @param module string
+--- @return integer, string
+local function import_module(prog, module)
+  local program = [[
+import sys, importlib.util;
+sys.path = [p for p in sys.path if p != ""];
+sys.stdout.write(str(sys.version_info[0]) + "." + str(sys.version_info[1]));]]
+
+  program = program
+    .. string.format('sys.exit(2 * int(importlib.util.find_spec("%s") is None))', module)
+
+  local out = vim.system({ prog, '-W', 'ignore', '-c', program }):wait()
+  return out.code, assert(out.stdout)
+end
+
+local function get_python_candidates()
+  return {
+    'python3',
+    'python3.12',
+    'python3.11',
+    'python3.10',
+    'python3.9',
+    'python3.8',
+    'python3.7',
+    'python',
+  }
+end
+
+--- @param prog string
+--- @param module string
+--- @param major_version integer
+--- @return integer, string
+local function check_for_module(prog, module, major_version)
+  local prog_path = vim.fn.exepath(prog)
+  if prog_path == '' then
+    return 0, prog .. ' not found in search path or not executable.'
+  end
+
+  local min_version = '3.7'
+
+  --   Try to load module, and output Python version.
+  --   Exit codes:
+  --     0  module can be loaded.
+  --     2  module cannot be loaded.
+  --     Otherwise something else went wrong (e.g. 1 or 127).
+  local prog_exitcode, prog_version = import_module(prog, module)
+  if prog_exitcode == 2 or prog_exitcode == 0 then
+    -- Check version only for expected return codes.
+    if not vim.startswith(prog_version, tostring(major_version)) then
+      return 0,
+        string.format(
+          '%s is Python %s and cannot provide Python %s.',
+          prog_path,
+          prog_version,
+          major_version
+        )
+    elseif
+      vim.startswith(prog_version, tostring(major_version))
+      and vim.version.lt(prog_version, min_version)
+    then
+      return 0,
+        string.format(
+          '%s is Python %s and cannot provide Python >= %s.',
+          prog_path,
+          prog_version,
+          min_version
+        )
+    end
+  end
+
+  if prog_exitcode == 2 then
+    return 0, string.format('%s does not have the "%s" module.', prog_path, module)
+  elseif prog_exitcode == 127 then
+    -- This can happen with pyenv's shims.
+    return 0, string.format('%s does not exist: %s', prog_path, prog_version)
+  elseif prog_exitcode ~= 0 then
+    return 0,
+      string.format(
+        'Checking %s caused an unknown error. (%s, output: %s) Report this at https://github.com/neovim/neovim',
+        prog_path,
+        prog_exitcode,
+        prog_version
+      )
+  end
+
+  return 1, ''
+end
+
+--- @param module string
+--- @param major_version integer
+--- @return string[]
+function M.detect_by_module(module, major_version)
+  if major_version ~= 3 then
```

Comment:
Similarly, such a "constant" could be defined in top level (and the check can use `vim.version()`).

Reply:
Hmm, good idea.